### PR TITLE
Adding Multi User & Role Tipping and Help/Warning Optimisation

### PR DESCRIPTION
**Pull request by ProfessorDey**
This includes my previous pull request. See individual commits for details

### CODE REVIEW

File: bot/modules/tipbot.js [hunk 1]
```diff
@@ -111,11 +166,86 @@ function doTip(message, tipper, words) {
...
+  }
+  var prv = 0;
+  var amountOffset = 2;
+  if (words.length >= 4 && words[1] === 'private') {
+    prv = 1;
+    amountOffset = 3;
+  }
+  let amount = getValidatedAmount(words[amountOffset]);
+  if (amount == null) {
+    message.reply('I dont know how to tip that many credits');
```
Comment (aayanl) [inline bot/modules/tipbot.js]: don't*

File: bot/modules/tipbot.js [hunk 2]
```diff
@@ -6,25 +6,88 @@ config = config.get('lbrycrd');
...
       words = msg.content.trim().split(' ').filter(function (n) { return n !== ""; }),
-      subcommand = words.length >= 2 ? words[1] : 'help';
+      subcommand = words.length >= 2 ? words[1] : 'help',
+      helpmsgparts = [['[help]', 'Get this message'],
+                      ['balance', 'Get your balance'],
+                      ['deposit', 'Get address for your deposits'],
+                      ['withdraw ADDRESS AMOUNT', 'Withdraw AMOUNT credits to ADDRESS'],
+                      ['[private] <user> <amount>', 'Mention a user with @ and then the amount to tip them, or put private before the user to tip them privately.']],
+      helpmsg = '```**!tip**\n' + formatDescriptions(helpmsgparts) + 'Key: [] : Optionally include contained keyword, <> : Replace with appropriate value.```',
+      channelwarning = 'Please use <#369896313082478594> or DMs to talk to bots.';
```
Comment (aayanl) [inline bot/modules/tipbot.js]: The channel shouldn't be hard-coded, rather in a config value somewhere

File: bot/modules/tipbot.js [hunk 3]
```diff
@@ -6,25 +6,88 @@ config = config.get('lbrycrd');
...
+  process: async function (bot, msg, suffix) {
+    let tipper = msg.author.id.replace('!', ''),
+        words = msg.content.trim().split(' ').filter(function (n) { return n !== ""; }),
+        subcommand = words.length >= 2 ? words[1] : 'help',
+        helpmsgparts = [['[help]', 'Get this message'],
+                        ['<user>+ <amount>', 'Mention one or more users in a row, seperated by spaces, then an amount that each mentioned user will receive.'],
+                        ['private <user>+ <amount>','Put private before the user list to have each user tipped privately, without revealing other users tipped.']],
+        helpmsg = '```**!multitip**\n' + formatDescriptions(helpmsgparts) + 'Key: [] : Optionally include contained keyword, <> : Replace with the appropriate value, + : Value can be repeated for multiple entries.```',
+        channelwarning = 'Please use <#369896313082478594> or DMs to talk to bots.';
```
Comment (aayanl) [inline bot/modules/tipbot.js]: ditto

**Comment by filipnyquist** [comment]
Looks good, i´ll test run this tomorrow on my sandbox server, i´ll assign myself this issue as nikoo catches up with life after studies :)

**Comment by ProfessorDey** [comment]
Code has been updated to include some code fixes from MSFTblockchain and also adds some additional functionality, while cleaning things up code wise.

**Comment by ProfessorDey** [comment]
And we are finally fully operational! Thanks a lot MSFTblockchain, you will be receiving half of whatever bounty or reward (personal or otherwise, Brazen likes his new custom features).

**Comment by MSFTserver** [comment]
roletip if user doesnt have an address assigned it fails to assign one and spams error 500

**Comment by ProfessorDey** [comment]
Unfortunately have no idea about the operation of the lbry backend because can't find any of that code anywhere. Also have no idea how new users are added as it's apparently supposed to be automatic when tipbot sends to a new address, meaning it _should_ be in lbry.sendFrom(), which I don't have access to.

**Comment by MSFTserver** [comment]
well this bot uses the bitcoind wallets and rpc commands if that helps

**Comment by MSFTserver** [comment]
NOTE: if you encounter some errors with sending transaction to too many people
i.e `Error: Invalid params, response status code: 500` 
if you check the debug log in your wallet folder youll notice lines say: 
`request rejected becasue http work queue depth exceeded, it can be increased with the -rpcworkqueue= setting`
simply add this to the wallet config 

| command | | description |
|:----------|-|:----------|
| rpcthreads=X | | Set the number of threads to service RPC calls (default: 4) |
| rpcworkqueue=XXX | | Sets the depth of the work queue to service RPC calls (default is 16) |
| rpcclienttimeout=X | | How many seconds bitcoin will wait for a complete RPC HTTP request |

**Comment by MSFTserver** [comment]
roletip and multitips private tipping mode isnt working

_Sent from my Galaxy Note5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by nikooo777** [comment]
as discussed with @ProfessorDey there are still a few things to be committed before this can be merged so I'll wait for him to give me to GO.

**Comment by ProfessorDey** [comment]
Ok @nikooo777 we're finally done, that should now be fully functional so long as you remember to install regex and set the rpc queue settings to higher values, since we have some large roles after all.

**Comment by nikooo777** [comment]
ok my changes have created conflicts. I'll see what i can do about it

**Comment by nikooo777** [comment]
Alright It took me a while but I think i managed to clean it up. Pushed to your fork so you can now test it again before I merge it.

**Comment by ProfessorDey** [comment]
As I've said, I can confirm this works fine on my end, so feel free to merge whenever.

**Comment by nikooo777** [comment]
Thank you for your contribution! /merged

**Pull request merged**